### PR TITLE
186 Auskunft über den Fortschritt von init AWerte über Rest anbieten

### LINKED ISSUE #186: Auskunft über den Stand der asynchronen Verarbeitung
# Akzeptanzkriterien
- [x] via get-Request kann der Stand abgerufen werden
- [x] Resource: asyncProgress

# Umsetzungshinweise
- Controller -> Service
- Security ist gepflegt

### PR DESCRIPTION
**Pull request by MrSebastian**
# Beschreibung:

Rest-Endpunkt zu AwerteProgress angeboten womit der Stand der Initialisierung der AWerte (Wahlberechtigte) abgerufen werden kann

### Backend ###
- [x] Doku aktualisiert - die bestehende Doku ist aus meiner Sicht fachlich ausreichend
- [x] Swagger-API vollständig
- [x] Unit-Tests gepflegt
- [X] Integrationstests gepflegt
- [X] Beispiel-Requests gepflegt

# Referenzen[^1]:

Verwandt mit Issue #

Closes #186 

> [^1]: _Nicht zutreffende Referenzen vor dem Speichern entfernen_

## Summary by CodeRabbit

- **New Features**
	- Introduced a new endpoint `/businessActions/asyncProgress` for retrieving asynchronous progress data.
	- Added a data transfer object `AsyncProgressDTO` to encapsulate progress-related information.

- **Bug Fixes**
	- Enhanced security testing for the new endpoint to ensure proper access control for anonymous and authenticated users.

- **Tests**
	- Added unit and integration tests for the `AsyncProgressController` and its associated mapping functionality.
	- Implemented tests to validate the mapping logic between `AsyncProgress` and `AsyncProgressDTO`.

- **Chores**
	- Created an HTTP request sequence for testing authentication and user information retrieval.

### CODE REVIEW

**Comment by MrSebastian** [comment]
@Coderabbitai review

File: wls-ergebnismeldung-service/src/test/resources/asyncprogress.http [hunk 1]
```diff
@@ -0,0 +1,22 @@
+### Get token wls_all
+POST {{ SSO_URL }}/auth/realms/wls_realm/protocol/openid-connect/token
+Content-Type: application/x-www-form-urlencoded
+
+password = test &
+grant_type = password &
+client_secret = top-secret &
+client_id = wls &
+username = wls_all
```
[the remark this answers is not in the document]
Comment (MrSebastian) [reply wls-ergebnismeldung-service/src/test/resources/asyncprogress.http:9]: http-Files in src/test/resources are for testing and will not be shipped with the application

File: wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressController.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import de.muenchen.oss.wahllokalsystem.ergebnismeldungservice.service.awerte.AsyncProgress;
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/businessActions/asyncProgress")
+@RequiredArgsConstructor
+public class AsyncProgressController {
```
[the remark this answers is not in the document]
Comment (MrSebastian) [reply wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressController.java:18]: is defined in SwaggerConfiguration

File: wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        @Test
+        void should_returnData_when_dataIsFound() {
+            val mockedAsyncProgressAsDTO = new AsyncProgressDTO(LocalDateTime.now(), LocalDateTime.now(), false, 0, 0, "");
+            Mockito.when(asyncProgressDTOMapper.toDTO(asyncProgress)).thenReturn(mockedAsyncProgressAsDTO);
+
+            val response = unitUnderTest.getAsyncProgress();
+
+            Assertions.assertThat(response).isEqualTo(mockedAsyncProgressAsDTO);
+        }
```
[the remark this answers is not in the document]
Comment (MrSebastian) [reply wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerTest.java:38]: we have the GlobalExceptionHandler class that should handle all non handled exceptions. Tests will be added in AsyncProgressControllerIntegrationTest

File: wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+            val response = api.perform(request)
+                    .andExpect(status().isOk())
+                    .andReturn().getResponse();
+            val asyncProgressDTO = objectMapper.readValue(response.getContentAsString(), AsyncProgressDTO.class);
+
+            val expectedAsyncProgress = asyncProgressDTOMapper.toDTO(asyncProgress);
+
+            Assertions.assertThat(asyncProgressDTO).usingRecursiveComparison().isEqualTo(expectedAsyncProgress);
+        }
```
[the remark this answers is not in the document]
Comment (MrSebastian) [reply wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerIntegrationTest.java:62]: only get is allowed. We dont test all possibles scenarios that we dont support.

File: wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressDTOMapper.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import org.mapstruct.Mapping;
+
+@Mapper
+public interface AsyncProgressDTOMapper {
+
+    @Mapping(target = "aWerteLoadingActive", source = "AWerteLoadingActive")
+    @Mapping(target = "aWerteTotal", source = "AWerteTotal")
+    @Mapping(target = "aWerteFinished", source = "AWerteFinished")
+    @Mapping(target = "aWerteNext", source = "AWerteNext")
+    AsyncProgressDTO toDTO(AsyncProgress asyncProgress);
```
[the remark this answers is not in the document]
Comment (MrSebastian) [reply wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressDTOMapper.java:14]: that is not possible. With lower case first letter the generation of the mapper class fails

File: wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrSebastian) [reply wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerTest.java:38]: test implemted with 0371f718f7ae6acf309a1bc920a16c2e8e5b3d07

File: wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressController.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+    private final AsyncProgressDTOMapper asyncProgressDTOMapper;
+    private final AsyncProgress asyncProgress;
+
+    @GetMapping
+    public AsyncProgressDTO getAsyncProgress() {
+        return asyncProgressDTOMapper.toDTO(asyncProgress);
+    }
```
[the remark this answers is not in the document]
Comment (MrSebastian) [reply wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressController.java:39]: description added with d21bc2c14b79b900dbb90d5b6662ba14f5ceb58b

File: wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
```
Comment (vjohnslhm) [inline wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/awerte/AsyncProgressControllerTest.java]:
```suggestion
```

**Review by vjohnslhm — CHANGES_REQUESTED** [review]
bitte pr beschreibung noch anpassen. der rest sieht gut aus

**Comment by MrSebastian** [comment]
PR-Beschreibung nachgetragen

**Review by vjohnslhm — APPROVED** [review]

**Pull request merged**